src/empirical.cpp

```diff
@@ -58,32 +65,43 @@ List get_empirical(List mmrm_data, NumericVector theta, NumericVector beta, Nume
-    matrix<double> residual_i = residual.block(start_i, 0, n_visits_i, 1);
+    matrix<double> sigma_inv = derivatives_by_group[subject_group_i]->get_inverse(visit_i, dist_i);
     matrix<double> gi_sqrt_root = G_sqrt.segment(start_i, n_visits_i).matrix().asDiagonal();
-    matrix<double> gi_simga_inv_chol = gi_sqrt_root * sigma_inv_chol;
-    matrix<double> xt_gi_simga_inv_chol = Xi.transpose() * gi_simga_inv_chol;
-    matrix<double> identity = matrix<double>::Identity(n_visits_i, n_visits_i);
-    if (jackknife) {
-      identity = identity - xt_gi_simga_inv_chol.transpose() * beta_vcov_matrix * xt_gi_simga_inv_chol;
-    }
-    matrix<double> xta = xt_gi_simga_inv_chol * identity.inverse();
-    matrix<double> z = xta * gi_simga_inv_chol.transpose() * residual_i;
-    meat = meat + z * z.transpose();
-    xt_g_simga_inv_chol.block(0, start_i, p, n_visits_i) = xt_gi_simga_inv_chol;
-    ax.block(start_i, 0, n_visits_i, p) = xta.transpose();
+    w.block(start_i, start_i, n_visits_i, n_visits_i) = gi_sqrt_root * sigma_inv * gi_sqrt_root;
+    chols.block(start_i, start_i, n_visits_i, n_visits_i) = derivatives_by_group[subject_group_i]->get_inverse_chol(visit_i, dist_i);
+    phi.block(start_i, start_i, n_visits_i, n_visits_i) = derivatives_by_group[subject_group_i]->get_sigma(visit_i, dist_i);
   }
-  matrix<double> h = xt_g_simga_inv_chol.transpose() * beta_vcov_matrix * xt_g_simga_inv_chol;
-  matrix<double> imh = matrix<double>::Identity(n_observations, n_observations) - h;
-  matrix<double> ax_xtx =  ax * beta_vcov_matrix;
+  std::chrono::steady_clock::time_point s4 = std::chrono::steady_clock::now();
+  matrix<double> weighted_phi = matrix<double>::Zero(n_observations, n_observations);
+  weighted_phi = (1 / G_sqrt).matrix().asDiagonal() * phi * (1 / G_sqrt).matrix().asDiagonal();
+  matrix<double> hxw = hx * w; // probably slow. use block calculation!
+  matrix<double> i_hxw = matrix<double>::Identity(n_observations, n_observations) - hxw;
   matrix<double> g = matrix<double>::Zero(n_observations, p * n_subjects);
+  std::chrono::steady_clock::time_point s5 = std::chrono::steady_clock::now();
   for (int i = 0; i < n_subjects; i++) {
     int start_i = subject_zero_inds[i];
     int n_visits_i = subject_n_visits[i];
-    g.block(0, i * p, n_observations, p) = imh.block(0, start_i, n_observations, n_visits_i) * ax_xtx.block(start_i, 0, n_visits_i, p);
+    matrix<double> ai = matrix<double>::Identity(n_visits_i, n_visits_i);
+    if (type == 1) { // 1 for jackknife
+      ai = (ai - hxw.block(start_i, start_i, n_visits_i, n_visits_i)).inverse();
+    } else if (type == 2) {
+      matrix<double> di = chols.block(start_i, start_i, n_visits_i, n_visits_i);
+      matrix<double> bi1 = di * i_hxw.block(start_i, 0, n_visits_i, n_observations);
+      matrix<double> bi = bi1 * phi * bi1.transpose();
+      matrix<double> bisq = matrix<double>(bi.completeOrthogonalDecomposition().pseudoInverse()).sqrt();
```

Comment:
it was used here but seems not very handy (sqrt is from eigen/unspported); this part returns some random nan but not sure why this happens